### ISSUE TITLE
Binding doesn't support coroutines

### ISSUE THREAD

**Issue by ditn** [michaelbull/kotlin-result#24]
I was excited to try out the `binding` block in our codebase but it doesn't support suspending functions. We currently use `Result` heavily with coroutines so this was a bit of a bummer. Is there any likelihood of this being added in the future? I'm not sure if it's out of scope for this project. I'm happy to raise a PR if you think it would be a worthwhile addition.

**Comment by michaelbull** [comment]
Can you elaborate as to what you're trying to code with a code example and what part isn't working?

**Comment by ditn** [comment]
Sure, so something like this, where `getDetails` and `hasPendingOrActive` are suspending functions:

```kotlin
launch {
    binding {
        val details = getDetails().bind()
        val hasPending = hasPendingOrActive().bind()

        details to hasPending
    }
}
```
This won't compile - `Suspension functions can be called only within a coroutine body`. The solution would be to create a version with a `suspend` lambda as a parameter:

```kotlin
inline fun <V, E> binding(crossinline suspend block: ResultBinding<E>.() -> V): Result<V, E>
```

But as I said, I don't know if this is a usecase that you necessarily want to support.

**Comment by Sottti** [comment]
I didn't try this myself but I was hoping for this to actually work in that scenario. Would be a great use case IMO! Definitely thumbs up to this issue.

**Comment by michaelbull** [comment]
@ditn can we add that variant with the `suspend` without defining an explicit dependency on kotlinx-coroutines?

Also cc: @Munzey for their opinion

**Comment by ditn** [comment]
Yes I believe so. You'll need the dependency for testing but that's it.

**Comment by Munzey** [comment]
sounds like a cool enhancement! 


would require slightly more modification than @ditn 's suggestion. we'd have to include a second `binding` function:

```kt
suspend inline fun <V, E> binding(crossinline block: suspend ResultBinding<E>.() -> V): Result<V, E> 
```

due to function naming ambiguity we'd then need to decide on either naming it something else  like `suspendBinding` or have a separate package like `com.github.michaelbull.result.coroutines`

would be curious to see how the benchmarking performs for this variant.

**Comment by ditn** [comment]
@Munzey yeah you're absolutely correct, I was being overly simplistic - you'll need a different method signature.

**Comment by michaelbull** [comment]
Okay, sounds like an acceptable approach to implement it with a different name. I think putting it in a different package with the same name is more likely to cause confusion. The name `suspendBinding` will do for now I think. If somebody is willing to make a PR with this and the appropriate tests (as mentioned, we are going to need a test-only depenedency on kotlinx-coroutines as a result), then we can go ahead and progress this.

**Comment by Munzey** [comment]
just an fyi, if someone does want to go ahead and open a pr for this, because its an mpp project, the coroutine dep you will need for testing is `kotlinx-coroutines-core-common`

**Comment by Munzey** [comment]
since nobody volunteered, i'm gonna take a look at putting a pr up for this today

**Comment by ditn** [comment]
Ah thank you. I was planning on taking a look this weekend or next week, but I suspect it'll be quicker if you do it.

**Comment by badvok666** [comment]
Similarly the ` andThen` function does not support passing in suspend functions. Such as `getData.andThen(::someSuppendFun)` is invalid. 
I'm still getting to grips with this lib and monad stuff in general so apologies if the query is misplaced.

**Comment by ditn** [comment]
IIRC you're having issues because you're using a method reference. We use `andThen` with suspending functions all the time.

**Comment by Munzey** [comment]
one thing that's bugging me - its a pain that the kotlin compiler can't disambiguate between the names if the suspend version of the function is named the same. it compiles the functions fine, youll just not be able to use them 😓 
the "clean" solution to me is to write an extension
```kt
suspend inline fun <V, E> CoroutineScope.binding(...) : Result<V, E>
```
but that would require pulling the kotlin coroutine common core into the main dependencies. @michaelbull would that be a deal breaker to you?
otherwise i will continue ahead with naming it `suspendBinding` (or i thought to keep the binding word front of mind, `bindingWithSuspend` or `bindingSuspend` as its already quite a long keyword)

**Comment by michaelbull** [comment]
@Munzey 

The [convention](https://www.youtube.com/watch?v=hQrFfwT1IMo&feature=youtu.be&t=2333) of making a function an extension on `CoroutineScope` is that it launches a new coroutine in some form (e.g. it calls `launch`). The binding function should be a `suspend fun` but shouldn't actually launch a coroutine itself, therefore making it an extension function on `CoroutineScope` would be breaking the convention that Roman tells us to follow in this presentation.

**Comment by Munzey** [comment]
ah good to know thanks @michaelbull

**Comment by badvok666** [comment]
@ditn Thanks, i see what you mean.

**Comment by ditn** [comment]
@badvok666 no problem, it's annoying that it's not supported but it's a Kotlin problem.

**Comment by Munzey** [comment]
opened a pr: https://github.com/michaelbull/kotlin-result/pull/25

haven't added benchmarking yet, see description.

**Comment by michaelbull** [comment]
Released in 1.1.8

**Comment by Munzey** [comment]
@badvok666 @ditn 
regarding
> IIRC you're having issues because you're using a method reference. We use andThen with suspending functions all the time.

It looks like when we bump this lib to 1.4 that will no longer be an issue, see section "Suspend conversion on callable references":

https://kotlinlang.org/docs/reference/whatsnew14.html#callable-reference-improvements
